src/cli/self_update/windows.rs

```diff
@@ -505,83 +512,111 @@ fn _apply_new_path(new_path: Option<Vec<u16>>) -> Result<()> {
-                Ok(Some(s))
-            } else {
-                warn!(
-                    "the registry key HKEY_CURRENT_USER\\Environment\\PATH is not a string. \
-                       Not modifying the PATH variable"
-                );
-                Ok(None)
-            }
+            // if let Some(s) = from_winreg_value(&val) {
+            //     Ok(Some(s))
+            // } else {
+            //     warn!(
+            //         "the registry key HKEY_CURRENT_USER\\Environment\\PATH is not a string. \
+            //            Not modifying the PATH variable"
+            //     );
+            //     Ok(None)
+            // }
+            Ok(Some(val))
+        }
+        Err(ref e) if e.code() == HRESULT::from_win32(ERROR_FILE_NOT_FOUND) => {
+            Ok(Some(String::new()))
         }
-        Err(ref e) if e.kind() == io::ErrorKind::NotFound => Ok(Some(Vec::new())),
         Err(e) => Err(e).context(CLIError::WindowsUninstallMadness),
     }
 }
 
 // Returns None if the existing old_path does not need changing, otherwise
 // prepends the path_str to old_path, handling empty old_path appropriately.
-fn _add_to_path(old_path: Vec<u16>, path_str: Vec<u16>) -> Option<Vec<u16>> {
+fn _add_to_path(old_path: String, path_str: String) -> Option<String> {
     if old_path.is_empty() {
         Some(path_str)
     } else if old_path
-        .windows(path_str.len())
-        .any(|path| path == path_str)
+        // .windows(path_str.len())
+        // .any(|path| path == path_str)
+        .contains(&path_str)
     {
         None
     } else {
         let mut new_path = path_str;
-        new_path.push(b';' as u16);
-        new_path.extend_from_slice(&old_path);
+        new_path.push(';');
+        new_path += &old_path;
         Some(new_path)
     }
 }
 
+// // Returns None if the existing old_path does not need changing
+// fn _remove_from_path(old_path: Vec<u16>, path_str: Vec<u16>) -> Option<Vec<u16>> {
+//     let idx = old_path
+//         .windows(path_str.len())
+//         .position(|path| path == path_str)?;
+//     // If there's a trailing semicolon (likely, since we probably added one
+//     // during install), include that in the substring to remove. We don't search
+//     // for that to find the string, because if it's the last string in the path,
+//     // there may not be.
+//     let mut len = path_str.len();
+//     if old_path.get(idx + path_str.len()) == Some(&(b';' as u16)) {
+//         len += 1;
+//     }
+
+//     let mut new_path = old_path[..idx].to_owned();
+//     new_path.extend_from_slice(&old_path[idx + len..]);
+//     // Don't leave a trailing ; though, we don't want an empty string in the
+//     // path.
+//     if new_path.last() == Some(&(b';' as u16)) {
+//         new_path.pop();
+//     }
+//     Some(new_path)
+// }
+
 // Returns None if the existing old_path does not need changing
-fn _remove_from_path(old_path: Vec<u16>, path_str: Vec<u16>) -> Option<Vec<u16>> {
```

Comment:
This implementation has some bugs, hence the test item `windows_uninstall_removes_semicolon_from_path_prefix` will be failed.

I don't know how to fix it.

Reply:
@InfyniteHeap This seems to be a regular Rust function that is not related to Windows-specific features. I encourage you to try out some input String examples with, say, emojis, or Chinese characters, to make sure they work correctly at least in the valid UTF8 case.

Reply:
I previously correctly implemented this function (sorry, I mis-referred to this deleted one, I must refer to the added one😅), but today I met some accidents so what my changed had been lost, and I forgot how did I implemented it, so I used this one.

Reply:
@InfyniteHeap No worries! Just don't forget to squash these commits with back-and-forth changes into one when you mark the PR as ready for review, we'd like to see atomic commits :)

Reply:
@InfyniteHeap Sorry for not being clear enough previously, you are allowed to have multiple commits per PR, and the goal of atomic commits is to make it clearer to see how you made it there step by step. For example, all function renaming actions can go into a separate commit, and replacing `winreg` calls with `windows-registry` calls can go into another one.

Reply:
> @InfyniteHeap Sorry for not being clear enough previously, but you are allowed to have multiple commits per PR. The goal of atomic commits is to make it clearer to see how you made it to the final state step by step. For example, replacing `winreg` calls with `windows-registry` calls can go into a separate commit, and all function renaming actions can go into another one.

So should I do something? Or just keep the current state?